test/e2e/annotations/rewrite.go

```diff
@@ -169,4 +175,106 @@ var _ = framework.IngressNginxDescribe("Annotations - Rewrite", func() {
+		_, err = f.EnsureIngress(ing)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(ing).NotTo(BeNil())
+
+		err = f.WaitForNginxServer(host,
+			func(server string) bool {
+				return strings.Contains(server, "location ~* ^/foo {")
+			})
+		Expect(err).NotTo(HaveOccurred())
+
+		err = f.WaitForNginxServer(host,
+			func(server string) bool {
+				return strings.Contains(server, "location ~* ^/foo.+ {")
+			})
+		Expect(err).NotTo(HaveOccurred())
+
+		By("ensuring '/foo' matches '~* ^/foo'")
+		expectBodyRequestURI := fmt.Sprintf("request_uri=http://%v:8080/foo", host)
+		resp, body, errs := gorequest.New().
+			Get(f.IngressController.HTTPURL+"/foo").
+			Set("Host", host).
+			End()
+
+		Expect(len(errs)).Should(Equal(0))
+		Expect(resp.StatusCode).Should(Equal(http.StatusOK))
+		Expect(body).Should(ContainSubstring(expectBodyRequestURI))
+
+		By("ensuring '/foo/bar' matches '~* ^/foo.+'")
+		expectBodyRequestURI = fmt.Sprintf("request_uri=http://%v:8080/foo/bar", host)
+		resp, body, errs = gorequest.New().
+			Get(f.IngressController.HTTPURL+"/foo/bar").
+			Set("Host", host).
+			End()
+
+		Expect(len(errs)).Should(Equal(0))
+		Expect(resp.StatusCode).Should(Equal(http.StatusOK))
+		Expect(body).Should(ContainSubstring(expectBodyRequestURI))
+	})
+
+	It("should fail to use longest match for documented edge case", func() {
+		host := "rewrite.bar.com"
+
+		By("creating a regular ingress definition")
+		ing := framework.NewSingleIngress("kube-lego", "/foo/bar/", host, f.IngressController.Namespace, "http-svc", 80, &map[string]string{})
+		_, err := f.EnsureIngress(ing)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(ing).NotTo(BeNil())
+
+		By(`creating an ingress definition with the use-regex annotation`)
+		annotations := map[string]string{
+			"nginx.ingress.kubernetes.io/use-regex":      "true",
+			"nginx.ingress.kubernetes.io/rewrite-target": "/new/backend",
+		}
+		ing = framework.NewSingleIngress("regex", "/foo/bar/[A-Z0-9]*", host, f.IngressController.Namespace, "http-svc", 80, &annotations)
+		_, err = f.EnsureIngress(ing)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(ing).NotTo(BeNil())
+
+		err = f.WaitForNginxServer(host,
+			func(server string) bool {
+				return strings.Contains(server, "location ~* ^/foo/bar/ {")
+			})
+		Expect(err).NotTo(HaveOccurred())
+
+		err = f.WaitForNginxServer(host,
+			func(server string) bool {
+				return strings.Contains(server, `location ~* ^/foo/bar/[A-Z0-9]*\/?(?<baseuri>.*) {`)
+			})
+		Expect(err).NotTo(HaveOccurred())
+
+		By("check that '/foo/bar/' matches the longest path")
+		expectBodyRequestURI := fmt.Sprintf("request_uri=http://%v:8080/new/backend", host)
+		resp, body, errs := gorequest.New().
+			Get(f.IngressController.HTTPURL+"/foo/bar/").
+			Set("Host", host).
+			End()
+
+		Expect(len(errs)).Should(Equal(0))
+		Expect(resp.StatusCode).Should(Equal(http.StatusOK))
+		Expect(body).Should(ContainSubstring(expectBodyRequestURI))
```

Comment:
My comment at https://github.com/kubernetes/ingress-nginx/pull/3145/files#r221124591 is not applicable here because of https://github.com/kubernetes/ingress-nginx/pull/3145/files#diff-780db26d3e78669defcf119bfb8a9b64R250, correct? Maybe you can do the same hack above.